### PR TITLE
[Clang] Automatically enable `-fconvergent-functions` on GPU targets

### PR DESCRIPTION
**Pull request by jhuber6**
Summary:
This patch causes us to respect the `-fconvergent-functions` and
`-fno-convergent-functions` options correctly. GPU targets should have
this set all the time, but we now offer `-fno-convergent-functions` to
opt-out if you want to test broken behavior. This munged about with a
lot of the old weird logic, but I don't think it makes any real changes.

### CODE REVIEW

**Comment by arsenm** [comment]
> -fno-convergent-functions to opt-out if you want to test broken behavior. 

You may legitimately know there are no convergent functions in the TU. We also have the noconvergent source attribute now for this

File: clang/lib/Frontend/CompilerInvocation.cpp [hunk 1]
```diff
@@ -4106,9 +4106,10 @@ bool CompilerInvocation::ParseLangArgs(LangOptions &Opts, ArgList &Args,
   Opts.Blocks = Args.hasArg(OPT_fblocks) || (Opts.OpenCL
     && Opts.OpenCLVersion == 200);
 
-  Opts.ConvergentFunctions = Args.hasArg(OPT_fconvergent_functions) ||
-                             Opts.OpenCL || (Opts.CUDA && Opts.CUDAIsDevice) ||
-                             Opts.SYCLIsDevice || Opts.HLSL;
+  Opts.ConvergentFunctions = Args.hasFlag(
+      OPT_fconvergent_functions, OPT_fno_convergent_functions,
+      Opts.OpenMPIsTargetDevice || T.isAMDGPU() || T.isNVPTX() || Opts.OpenCL ||
+          Opts.CUDAIsDevice || Opts.SYCLIsDevice || Opts.HLSL);
```
Comment (arsenm) [inline clang/lib/Frontend/CompilerInvocation.cpp:4112]: Sort all the language checks together, before the target list. We probably should have a hasConvergentOperations() predicate somewhere

Comment (jhuber6) [reply clang/lib/Frontend/CompilerInvocation.cpp]: Done, I just put it in a variable for now.

**Review by arsenm — APPROVED** [review]

**Comment by jhuber6** [comment]
> > -fno-convergent-functions to opt-out if you want to test broken behavior.
> 
> You may legitimately know there are no convergent functions in the TU. We also have the noconvergent source attribute now for this

Think it would be useful to put that on functions in the wrapper headers that definitely aren't convergent? E.g. getting a thread id.

**Comment by arsenm** [comment]
> Think it would be useful to put that on functions in the wrapper headers that definitely aren't convergent? E.g. getting a thread id.

You could, but it's trivially inferable in those cases anyway

**Pull request merged**